# Data Section

In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

In [9]:
!pip install googlemaps
import googlemaps
import json
gmaps = googlemaps.Client(key='AIzaSyBHPgUusU8uUrhtl6I7l7SqSPzKZi2O7NA')

Requirement not upgraded as not directly required: googlemaps in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages
Requirement not upgraded as not directly required: requests<3.0,>=2.11.1 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from googlemaps)
Requirement not upgraded as not directly required: chardet<3.1.0,>=3.0.2 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests<3.0,>=2.11.1->googlemaps)
Requirement not upgraded as not directly required: idna<2.7,>=2.5 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests<3.0,>=2.11.1->googlemaps)
Requirement not upgraded as not directly required: urllib3<1.23,>=1.21.1 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests<3.0,>=2.11.1->googlemaps)
Requirement not upgraded as not directly required: certifi>=2017.4.17 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests<3.0,>=2.11.1->googlemaps)


In [12]:
def get_pin(page,Area_Name,Pin_Code) :
    wiki_link='http://www.chennaiiq.com/india/pincode/index.asp?id=35&state_name=West%20Bengal&page='+page
    raw_wiki_page_content = requests.get(wiki_link).text
#print(wiki_link)
    raw_wiki_page_content_xml = BeautifulSoup(raw_wiki_page_content,'html.parser')
    table = None
    for tab_cell in raw_wiki_page_content_xml.find_all('table', class_='TBox'):
        table = tab_cell
    line_number = 1
    for tr_cell in table.find_all('tr'):
        if line_number > 1 :
            column_number = 1
            for td_cell in tr_cell.find_all('td'):
                if column_number == 3 :
                #print(td_cell.text)
                    Area_Name.append(td_cell.text)
                if column_number == 5 :
                    Pin_Code.append(td_cell.text)
                column_number = column_number + 1
        line_number = line_number + 1


In [13]:
page_list = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S',
'T','U','V','W','X','Y','Z']
Area_Name = []
Pin_Code = []
for p in page_list:
    get_pin(p,Area_Name,Pin_Code)
    kolkata_df = pd.DataFrame({"AreaName":Area_Name,"PinCode":Pin_Code})
kolkata_df

AreaName PinCode
0       Ananda Bazar Patrika  700001
1               Amrita Bazar  700003
2                  Ahiritola  700005
3                    Archana  700007
4          Archarya Jagadish  700020
5     Alipore Bodyguard Line  700027
6        Alipore Civil Court  700027
7     Alipore District Board  700027
8                Alipore H O  700027
9         Alipore Store Yard  700027
10                 Alambazar  700035
11                Ashokegarh  700035
12    Abinash Choudhury Lane  700046
13                  Ariadaha  700057
14                  Agarpara  700109
15                  Agarpara  700109
16                Anandapuri  700122
17                  Abdalpur  700155
18                Andul Road  711103
19       A Road (Bamongachi)  711106
20           Abhoy Guha Road  711204
21                 Anantapur  711301
22               Andul Mauri  711302
23                      Amta  711401
24                 Adconagar  712121
25             Ayemanawabpur  712155
26                    Antpar  712424
27            Arambagh Court  712601
28              Arambagh H O  712601
29         Agricultural Firm  713101
...                      ...     ...
2042                Uluberia  711315
2043               Uttarpara  712258
2044         Uttarpara Bazar  712258
2045            Uniatatarpur  713173
2046                Ushagram  713303
2047       Upper Chellidanga  713304
2048   Ukhra Puratan Hattala  713363
2049                   Ukhra  713363
2050                 Uchalon  713427
2051                     Ura  713433
2052       Uttar Soutan Chak  721664
2053                Ukilpara  741101
2054                   Usthi  743375
2055            Uttar Raipur  743389
2056             Vishyasabha  700007
2057            Viviekanagar  700075
2058               Vip Nagar  700100
2059        Vivekananda Math  700120
2060       Vivekananda Pally  700139
2061       Vivekananda Pally  700139
2062               V C Mills  711107
2063               V K Nagar  713210
2064          Vivekandanagar  723147
2065      Vidyalaya Mohanpur  741252
2066            W B Assembly  700001
2067        Writers Building  700001
2068                Watgunge  700023
2069            W B G S Camp  700062
2070       West Congresspara  735101
2071             Zone Colony  713201

[2072 rows x 2 columns]

In [14]:
kolkata_df = kolkata_df.groupby('PinCode').agg({'AreaName': ', '.join}).reset_index()
kolkata_df

PinCode                                           AreaName
0     700001  Ananda Bazar Patrika, Braborne Road, Council H...
1     700002  Cossipore Gun Factory, Cossipore H O, Indian R...
2     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika
3     700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...
4     700005                                Ahiritola, Hatkhola
5     700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...
6     700007  Archana, Barabazar H O, College Squire, Kalake...
7     700008                             Barisha, Purba Barisha
8     700009                                  Sarani Rarsibagan
9     700010          Beleghata H O, Joranandir, Subhas Sarobar
10    700011                                        Narkeldanga
11    700012                             Bow Bazar, Sankaritala
12    700013                              Dharamtala, Mott Lane
13    700014  Bamboovilla, Intally, Linton Street, Sealdah, ...
14    700015              Seal Lane, Sales Tax Building, Tengra
15    700016             Elliot Road, Madrassa, Park Street H O
16    700017  Circus Avenue, Jhowtala, Park Circus, Shakespe...
17    700018                             Bartala, Rabindranagar
18    700019  Ballygunge, Ballygunge R S, Ballygunge Science...
19    700020  Archarya Jagadish, Chandra Bose Road, Gokhale ...
20    700021                                       Fort Willium
21    700022                              Bakery Road, Hastings
22    700023  Kidderpore, Manstala, Market, Raja Janakinath,...
23    700024  Garden Reach, Keshoram Cotton Mills, Paharpur ...
24    700025  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...
25    700026  Keoratala, Kalighat, Kalimandir, Protisthan, R...
26    700027  Alipore Bodyguard Line, Alipore Civil Court, A...
27    700028  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...
28    700029  Dover Lane, Lake Market, Rash Behari Avenue, S...
29    700030              Ghughudanga, Purbe Sinthee, Sethbagan
...      ...                                                ...
1355  743440                                              Metia
1356  743441                                           Minakhan
1357  743442                                       Nazathtkhola
1358  743443                                            Hadipur
1359  743501                                 Achipur, Bansdroni
1360  743502                                            Bhangar
1361  743503                                 Bishnupur Baruipur
1362  743504                                              Falta
1363  743507                                          Jairampur
1364  743512                                         I I M Joka
1365  743513                                           Fathepur
1366  743514                                           Chaberia
1367  743517                                              Bodra
1368  743601                                           Gokarnee
1369  743602                                         Rangabelia
1370  743603                                            Kautala
1371  743604                                          Gilarchat
1372  743605                                              Nalua
1373  743606                                         Gangasagar
1374  743607                                          Jalaberia
1375  743608                                   Bakultala 24 Pgs
1376  743609                                            Hamsuri
1377  743610                                            Madarat
1378  743611                                       Pathan Khali
1379  743613                                        South Garia
1380  743701                                              Palla
1381  743702                                           Rajibpur
1382  743703                                           Tentulia
1383  743704                                               Guma
1384  743705                                          Petrapole

[1385 rows x 2 columns]

In [15]:
kolkata_df.to_csv('wb_area_to_pin.csv')

In [16]:
Area_Name = []
Pin_Code = []
Lat = []
Lng = []
for index, row in kolkata_df.iterrows():
    PinCode_temp = row['PinCode']
    PinCode_temp = PinCode_temp.strip()
    AreaName_temp = row['AreaName']
    AreaName_temp = AreaName_temp.strip()
    Pin_Code.append(PinCode_temp)
    Area_Name.append(AreaName_temp)
    try:
        geocode_result = gmaps.geocode(PinCode_temp+', West Bengal, India')
        geocode_result_s = str(geocode_result)
        geocode_result_s = geocode_result_s.replace("'","\"")
        geocode_result_s = geocode_result_s[1:len(geocode_result_s)-1]
        #print(geocode_result_s)
        geocode_result_j = json.loads(geocode_result_s)
        lat = geocode_result_j["geometry"]["location"]["lat"]
        lng = geocode_result_j["geometry"]["location"]["lng"]
        Lat.append(lat)
        Lng.append(lng)
        print("PinCode_temp:"+PinCode_temp+" lat:"+str(lat))
    except:
        Lat.append(-1)
        Lng.append(-1)
        print("PinCode_temp:"+PinCode_temp+" lat:"+str(-1))

PinCode_temp:700001 lat:22.5736132
PinCode_temp:700002 lat:22.6189402
PinCode_temp:700003 lat:22.6038219
PinCode_temp:700004 lat:22.5983206
PinCode_temp:700005 lat:22.5980678
PinCode_temp:700006 lat:22.5890726
PinCode_temp:700007 lat:22.5825574
PinCode_temp:700008 lat:22.4811681
PinCode_temp:700009 lat:22.5766749
PinCode_temp:700010 lat:22.5630255
PinCode_temp:700011 lat:22.577486
PinCode_temp:700012 lat:22.5683534
PinCode_temp:700013 lat:22.5653004
PinCode_temp:700014 lat:22.5596244
PinCode_temp:700015 lat:22.5550397
PinCode_temp:700016 lat:22.5535048
PinCode_temp:700017 lat:22.5409396
PinCode_temp:700018 lat:22.5360345
PinCode_temp:700019 lat:22.5279995
PinCode_temp:700020 lat:22.5377634
PinCode_temp:700021 lat:22.5636078
PinCode_temp:700022 lat:22.5482461
PinCode_temp:700023 lat:22.5317391
PinCode_temp:700024 lat:22.5484801
PinCode_temp:700025 lat:22.5301479
PinCode_temp:700026 lat:22.5187493
PinCode_temp:700027 lat:22.5259613
PinCode_temp:700028 lat:22.6319864
PinCode_temp:700029 l

In [18]:
kolkata_df = pd.DataFrame({"AreaName":Area_Name,"PinCode":Pin_Code,"Latitude":Lat,"Longitude":Lng})
kolkata_df

AreaName   Latitude  Longitude  \
0     Ananda Bazar Patrika, Braborne Road, Council H...  22.573613  88.348341   
1     Cossipore Gun Factory, Cossipore H O, Indian R...  22.618940  88.373169   
2      Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822  88.367960   
3     R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321  88.374501   
4                                   Ahiritola, Hatkhola  22.598068  88.358119   
5     Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073  88.368755   
6     Archana, Barabazar H O, College Squire, Kalake...  22.582557  88.361703   
7                                Barisha, Purba Barisha  22.481168  88.313455   
8                                     Sarani Rarsibagan  22.576675  88.373545   
9             Beleghata H O, Joranandir, Subhas Sarobar  22.563025  88.396257   
10                                          Narkeldanga  22.577486  88.380986   
11                               Bow Bazar, Sankaritala  22.568353  88.361571   
12                                Dharamtala, Mott Lane  22.565300  88.354289   
13    Bamboovilla, Intally, Linton Street, Sealdah, ...  22.559624  88.370501   
14                Seal Lane, Sales Tax Building, Tengra  22.555040  88.381247   
15               Elliot Road, Madrassa, Park Street H O  22.553505  88.356574   
16    Circus Avenue, Jhowtala, Park Circus, Shakespe...  22.540940  88.368135   
17                               Bartala, Rabindranagar  22.536034  88.263908   
18    Ballygunge, Ballygunge R S, Ballygunge Science...  22.527999  88.366045   
19    Archarya Jagadish, Chandra Bose Road, Gokhale ...  22.537763  88.351468   
20                                         Fort Willium  22.563608  88.338588   
21                                Bakery Road, Hastings  22.548246  88.329701   
22    Kidderpore, Manstala, Market, Raja Janakinath,...  22.531739  88.319618   
23    Garden Reach, Keshoram Cotton Mills, Paharpur ...  22.548480  88.287661   
24    Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...  22.530148  88.346228   
25    Keoratala, Kalighat, Kalimandir, Protisthan, R...  22.518749  88.342573   
26    Alipore Bodyguard Line, Alipore Civil Court, A...  22.525961  88.334114   
27    Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...  22.631986  88.416555   
28    Dover Lane, Lake Market, Rash Behari Avenue, S...  22.517586  88.349433   
29                Ghughudanga, Purbe Sinthee, Sethbagan  22.620038  88.393611   
...                                                 ...        ...        ...   
1355                                              Metia  -1.000000  -1.000000   
1356                                           Minakhan  -1.000000  -1.000000   
1357                                       Nazathtkhola  22.448604  88.807089   
1358                                            Hadipur  -1.000000  -1.000000   
1359                                 Achipur, Bansdroni  -1.000000  -1.000000   
1360                                            Bhangar  22.494411  88.575857   
1361                                 Bishnupur Baruipur  22.360813  88.314212   
1362                                              Falta  22.299429  88.148657   
1363                                          Jairampur  22.444075  88.612071   
1364                                         I I M Joka  -1.000000  -1.000000   
1365                                           Fathepur  22.312735  88.275314   
1366                                           Chaberia  -1.000000  -1.000000   
1367                                              Bodra  -1.000000  -1.000000   
1368                                           Gokarnee  22.269304  88.399179   
1369                                         Rangabelia  -1.000000  -1.000000   
1370                                            Kautala  22.045793  88.396754   
1371                                          Gilarchat  -1.000000  -1.000000   
1372                                              Nalua  -1.000000  -1.000000 

In [19]:
#removing areas with undefined latitude/longitude
kolkata_df = kolkata_df[kolkata_df.Latitude != -1]
kolkata_df

AreaName   Latitude  Longitude  \
0     Ananda Bazar Patrika, Braborne Road, Council H...  22.573613  88.348341   
1     Cossipore Gun Factory, Cossipore H O, Indian R...  22.618940  88.373169   
2      Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822  88.367960   
3     R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321  88.374501   
4                                   Ahiritola, Hatkhola  22.598068  88.358119   
5     Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073  88.368755   
6     Archana, Barabazar H O, College Squire, Kalake...  22.582557  88.361703   
7                                Barisha, Purba Barisha  22.481168  88.313455   
8                                     Sarani Rarsibagan  22.576675  88.373545   
9             Beleghata H O, Joranandir, Subhas Sarobar  22.563025  88.396257   
10                                          Narkeldanga  22.577486  88.380986   
11                               Bow Bazar, Sankaritala  22.568353  88.361571   
12                                Dharamtala, Mott Lane  22.565300  88.354289   
13    Bamboovilla, Intally, Linton Street, Sealdah, ...  22.559624  88.370501   
14                Seal Lane, Sales Tax Building, Tengra  22.555040  88.381247   
15               Elliot Road, Madrassa, Park Street H O  22.553505  88.356574   
16    Circus Avenue, Jhowtala, Park Circus, Shakespe...  22.540940  88.368135   
17                               Bartala, Rabindranagar  22.536034  88.263908   
18    Ballygunge, Ballygunge R S, Ballygunge Science...  22.527999  88.366045   
19    Archarya Jagadish, Chandra Bose Road, Gokhale ...  22.537763  88.351468   
20                                         Fort Willium  22.563608  88.338588   
21                                Bakery Road, Hastings  22.548246  88.329701   
22    Kidderpore, Manstala, Market, Raja Janakinath,...  22.531739  88.319618   
23    Garden Reach, Keshoram Cotton Mills, Paharpur ...  22.548480  88.287661   
24    Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...  22.530148  88.346228   
25    Keoratala, Kalighat, Kalimandir, Protisthan, R...  22.518749  88.342573   
26    Alipore Bodyguard Line, Alipore Civil Court, A...  22.525961  88.334114   
27    Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...  22.631986  88.416555   
28    Dover Lane, Lake Market, Rash Behari Avenue, S...  22.517586  88.349433   
29                Ghughudanga, Purbe Sinthee, Sethbagan  22.620038  88.393611   
...                                                 ...        ...        ...   
1341                                       Basirhat R S  22.655834  88.856115   
1342                                              Hebia  22.609947  88.869495   
1343                                            Deganga  22.685414  88.614484   
1344                                           Devalaya  22.733759  88.672686   
1345                                              Haria  22.543059  88.691631   
1346                                           Hansabad  22.532777  88.912645   
1347                                           Katiahat  22.734901  88.845504   
1348                                          Kholapota  22.675111  88.816696   
1349                                  Taki Market, Taki  22.604284  88.912645   
1350                                         Hinglaganj  22.449660  88.960536   
1351                                             Bhabla  22.554386  88.822841   
1352                                        Dhanyakuria  22.692588  88.763829   
1353                                             Kolsur  22.793834  88.759020   
1354                                          Hatgachha  22.247199  88.989245   
1357                                       Nazathtkhola  22.448604  88.807089   
1360                                            Bhangar  22.494411  88.575857   
1361                                 Bishnupur Baruipur  22.360813  88.314212   
1362                                              Falta  22.299429  88.148657 

In [28]:
# calculate distance from kolkata for all the areas
import geopy.distance

kol_latitude = 22.5726; kol_longitude = 88.3639 #From Google search
Area_Name = []
Pin_Code = []
Lat = []
Lng = []
Distance = []

for index, row in kolkata_df.iterrows():
    PinCode_temp = row['PinCode']
    AreaName_temp = row['AreaName']
    AreaName_temp = AreaName_temp.strip()
    Lat_temp = row['Latitude']
    Lng_temp = row['Longitude']
    distance = geopy.distance.vincenty((kol_latitude,kol_longitude), (Lat_temp,Lng_temp)).km
    Area_Name.append(AreaName_temp)
    Pin_Code.append(PinCode_temp)
    Lat.append(Lat_temp)
    Lng.append(Lng_temp)
    Distance.append(distance)

kolkata_df = pd.DataFrame({"PinCode":Pin_Code,"AreaName":Area_Name,"Latitude":Lat,"Longitude":Lng,"Distance":Distance})
kolkata_df

AreaName    Distance  \
0     Ananda Bazar Patrika, Braborne Road, Council H...    1.604059   
1     Cossipore Gun Factory, Cossipore H O, Indian R...    5.219402   
2      Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika    3.482579   
3     R G Kar M C Hospital, Shyambazar, Shyambazar M...    3.049759   
4                                   Ahiritola, Hatkhola    2.882242   
5     Beadon Street, Bidhan Sarani, Jorasanko, Manik...    1.891232   
6     Archana, Barabazar H O, College Squire, Kalake...    1.125577   
7                                Barisha, Purba Barisha   11.377439   
8                                     Sarani Rarsibagan    1.089683   
9             Beleghata H O, Joranandir, Subhas Sarobar    3.492611   
10                                          Narkeldanga    1.838534   
11                               Bow Bazar, Sankaritala    0.527732   
12                                Dharamtala, Mott Lane    1.276920   
13    Bamboovilla, Intally, Linton Street, Sealdah, ...    1.589218   
14                Seal Lane, Sales Tax Building, Tengra    2.639025   
15               Elliot Road, Madrassa, Park Street H O    2.244778   
16    Circus Avenue, Jhowtala, Park Circus, Shakespe...    3.532956   
17                               Bartala, Rabindranagar   11.053161   
18    Ballygunge, Ballygunge R S, Ballygunge Science...    4.943883   
19    Archarya Jagadish, Chandra Bose Road, Gokhale ...    4.064120   
20                                         Fort Willium    2.787156   
21                                Bakery Road, Hastings    4.432292   
22    Kidderpore, Manstala, Market, Raja Janakinath,...    6.420265   
23    Garden Reach, Keshoram Cotton Mills, Paharpur ...    8.283708   
24    Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...    5.040230   
25    Keoratala, Kalighat, Kalimandir, Protisthan, R...    6.354008   
26    Alipore Bodyguard Line, Alipore Civil Court, A...    6.005020   
27    Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...    8.518210   
28    Dover Lane, Lake Market, Rash Behari Avenue, S...    6.271224   
29                Ghughudanga, Purbe Sinthee, Sethbagan    6.076949   
...                                                 ...         ...   
1109                                       Basirhat R S   51.437930   
1110                                              Hebia   52.153851   
1111                                            Deganga   28.629683   
1112                                           Devalaya   36.411379   
1113                                              Haria   33.866670   
1114                                           Hansabad   56.614367   
1115                                           Katiahat   52.662186   
1116                                          Kholapota   47.913570   
1117                                  Taki Market, Taki   56.536775   
1118                                         Hinglaganj   62.878204   
1119                                             Bhabla   47.244761   
1120                                        Dhanyakuria   43.205802   
1121                                             Kolsur   47.421366   
1122                                          Hatgachha   73.784188   
1123                                       Nazathtkhola   47.621437   
1124                                            Bhangar   23.460564   
1125                                 Bishnupur Baruipur   24.003637   
1126                                              Falta   37.497005   
1127                                          Jairampur   29.232959   
1128                                           Fathepur   30.186730   
1129                                           Gokarnee   33.781633   
1130                                            Kautala   58.433863   
1131                                         Gangasagar  106.223651   
1132                                            Hamsuri   31.346782   
1133                                            Madarat   29.036434   


In [29]:
kolkata_df.to_csv('wb_area_pin_lat_lng_dist.csv')

In [30]:
kolkata_20_df = kolkata_df[kolkata_df.Distance <= 20]
kolkata_20_df = kolkata_20_df.reset_index()
kolkata_20_df

index                                           AreaName   Distance  \
0        0  Ananda Bazar Patrika, Braborne Road, Council H...   1.604059   
1        1  Cossipore Gun Factory, Cossipore H O, Indian R...   5.219402   
2        2   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   3.482579   
3        3  R G Kar M C Hospital, Shyambazar, Shyambazar M...   3.049759   
4        4                                Ahiritola, Hatkhola   2.882242   
5        5  Beadon Street, Bidhan Sarani, Jorasanko, Manik...   1.891232   
6        6  Archana, Barabazar H O, College Squire, Kalake...   1.125577   
7        7                             Barisha, Purba Barisha  11.377439   
8        8                                  Sarani Rarsibagan   1.089683   
9        9          Beleghata H O, Joranandir, Subhas Sarobar   3.492611   
10      10                                        Narkeldanga   1.838534   
11      11                             Bow Bazar, Sankaritala   0.527732   
12      12                              Dharamtala, Mott Lane   1.276920   
13      13  Bamboovilla, Intally, Linton Street, Sealdah, ...   1.589218   
14      14              Seal Lane, Sales Tax Building, Tengra   2.639025   
15      15             Elliot Road, Madrassa, Park Street H O   2.244778   
16      16  Circus Avenue, Jhowtala, Park Circus, Shakespe...   3.532956   
17      17                             Bartala, Rabindranagar  11.053161   
18      18  Ballygunge, Ballygunge R S, Ballygunge Science...   4.943883   
19      19  Archarya Jagadish, Chandra Bose Road, Gokhale ...   4.064120   
20      20                                       Fort Willium   2.787156   
21      21                              Bakery Road, Hastings   4.432292   
22      22  Kidderpore, Manstala, Market, Raja Janakinath,...   6.420265   
23      23  Garden Reach, Keshoram Cotton Mills, Paharpur ...   8.283708   
24      24  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...   5.040230   
25      25  Keoratala, Kalighat, Kalimandir, Protisthan, R...   6.354008   
26      26  Alipore Bodyguard Line, Alipore Civil Court, A...   6.005020   
27      27  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...   8.518210   
28      28  Dover Lane, Lake Market, Rash Behari Avenue, S...   6.271224   
29      29              Ghughudanga, Purbe Sinthee, Sethbagan   6.076949   
..     ...                                                ...        ...   
156    167                                      Sambay Pallay   9.563899   
157    171                                          Ghoshpara   8.857075   
158    173                                        Andul Mauri  19.127230   
159    180                                        Delta Mills  14.315007   
160    182                               Jagaeha, Old Jagacha  15.722860   
161    184                                           Sankrail  15.722860   
162    191                                          Baltikuri   9.898265   
163    192                                               Bank   9.428650   
164    194                                             Domjur  17.148003   
165    196                                           Makardah  12.275647   
166    198                                              Begri  19.221497   
167    224  Kalbazar, Manicktala Bazar, Serampore Court, S...  19.538008   
168    225                                            Mahes-I  18.021132   
169    226                                        Mallickpara  19.414893   
170    231                                         Bhadrakali  11.310315   
171    232                                         Hind Motor  12.679277   
172    233                                           Kanaipur  14.503766   
173    234                              Cripar Road, Konnagar  14.285127   
174    235                                             Makhla  11.576381   
175    236                                 Nabagram (Hooghly)  14.142421   
176    237                             Raghunathpur (

In [34]:
!pip install folium
import folium

kol_latitude = 22.5726; kol_longitude = 88.3639 #From Google search
map_kolkata = folium.Map(location = [kol_latitude, kol_longitude], zoom_start = 10.7)

# add markers to map
for lat, lng, pincode, area in zip(kolkata_20_df['Latitude'], kolkata_20_df['Longitude'], kolkata_20_df['PinCode'], kolkata_20_df['AreaName']):
    label = '{}, {}'.format(pincode, area)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_kolkata)
map_kolkata

    100% |████████████████████████████████| 92kB 7.5MB/s eta 0:00:01
Requirement not upgraded as not directly required: jinja2 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from folium)
Requirement not upgraded as not directly required: six in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from folium)
Requirement not upgraded as not directly required: requests in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from folium)
Requirement not upgraded as not directly required: numpy in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from folium)
Requirement not upgraded as not directly required: MarkupSafe>=0.23 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from jinja2->folium)
Requirement not upgraded as not directly required: chardet<3.1.0,>=3.0.2 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests->folium)
Requirement not upgraded as not directly required: idna<2.7,>=2.5 in /opt/conda/envs/DSX-Python35/

In [35]:
# Populate relevant venues nearby all the pincodes selected
CLIENT_ID = 'J1JBTE1UXQVN0P0W2TV04LA2HMJ3ZYMOY0QOA42IWYCIG0CW'
CLIENT_SECRET = 'XWM0RWPJN11YUG34VHB1QQI0FSPBJOYE3AJ2RAHCQTIWRDUZ'
VERSION = '20190308'
LIMIT = 100
radius = 500

In [36]:
import json # library to handle JSON files
from pandas.io.json import json_normalize
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [37]:
Pin_Code = []
Area_Name = []
Lat = []
Lng = []
Distance = []
Venue_name = []
Venue_cat = []
Venue_lat = []
Venue_lng = []

for index, row in kolkata_20_df.iterrows():
    PinCode_temp = row['PinCode']
    AreaName_temp = row['AreaName']
    Lat_temp = row['Latitude']
    Lng_temp = row['Longitude']
    dist_temp = row['Distance']

    try:
        url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, Lat_temp, Lng_temp, VERSION, radius, LIMIT)
        results = requests.get(url).json()
        venues = results['response']['groups'][0]['items']
        nearby_venues = json_normalize(venues)
        nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
        
        # clean columns
        nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
        for index1, row1 in nearby_venues.iterrows():
            Pin_Code.append(PinCode_temp)
            Area_Name.append(AreaName_temp)
            Lat.append(Lat_temp)
            Lng.append(Lng_temp)
            Distance.append(dist_temp)
            name_temp = row1['name']
            categories_temp = row1['categories']
            lat1_temp = row1['lat']
            lng1_temp = row1['lng']
            Venue_name.append(name_temp)
            Venue_cat.append(categories_temp)
            Venue_lat.append(lat1_temp)
            Venue_lng.append(lng1_temp)
            
    except:
        Pin_Code.append(PinCode_temp)
        Area_Name.append(AreaName_temp)
        Lat.append(Lat_temp)
        Lng.append(Lng_temp)
        Distance.append(dist_temp)
        name_temp = row1['name']
        Venue_name.append('None')
        Venue_cat.append('None')
        Venue_lat.append(-1)
        Venue_lng.append(-1)
        
kolkata_20_df = pd.DataFrame({"PinCode":Pin_Code,"AreaName":Area_Name,"Latitude":Lat,"Longitude":Lng,"Distance":Distance,"VenueName":Venue_name,"Category":Venue_cat,"VenueLatitude":Venue_lat,"VenueLongitude":Venue_lng})
kolkata_20_df

AreaName  \
0    Ananda Bazar Patrika, Braborne Road, Council H...   
1    Ananda Bazar Patrika, Braborne Road, Council H...   
2    Cossipore Gun Factory, Cossipore H O, Indian R...   
3     Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   
4     Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   
5     Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   
6     Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   
7     Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   
8     Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika   
9    R G Kar M C Hospital, Shyambazar, Shyambazar M...   
10   R G Kar M C Hospital, Shyambazar, Shyambazar M...   
11   R G Kar M C Hospital, Shyambazar, Shyambazar M...   
12   R G Kar M C Hospital, Shyambazar, Shyambazar M...   
13   R G Kar M C Hospital, Shyambazar, Shyambazar M...   
14   R G Kar M C Hospital, Shyambazar, Shyambazar M...   
15                                 Ahiritola, Hatkhola   
16   Beadon Street, Bidhan Sarani, Jorasanko, Manik...   
17   Beadon Street, Bidhan Sarani, Jorasanko, Manik...   
18   Beadon Street, Bidhan Sarani, Jorasanko, Manik...   
19   Beadon Street, Bidhan Sarani, Jorasanko, Manik...   
20   Archana, Barabazar H O, College Squire, Kalake...   
21   Archana, Barabazar H O, College Squire, Kalake...   
22   Archana, Barabazar H O, College Squire, Kalake...   
23   Archana, Barabazar H O, College Squire, Kalake...   
24   Archana, Barabazar H O, College Squire, Kalake...   
25                              Barisha, Purba Barisha   
26                              Barisha, Purba Barisha   
27                                   Sarani Rarsibagan   
28           Beleghata H O, Joranandir, Subhas Sarobar   
29           Beleghata H O, Joranandir, Subhas Sarobar   
..                                                 ...   
562                                            Mahes-I   
563                                        Mallickpara   
564                                         Bhadrakali   
565                                         Hind Motor   
566                                           Kanaipur   
567                              Cripar Road, Konnagar   
568                              Cripar Road, Konnagar   
569                                             Makhla   
570                                 Nabagram (Hooghly)   
571                             Raghunathpur (Hooghly)   
572  Bangurpark, Gandhisarak, Mahesh-Ii, Masjidlane...   
573        Raiendra Avenue, Uttarpara, Uttarpara Bazar   
574        Raiendra Avenue, Uttarpara, Uttarpara Bazar   
575                                              Janai   
576                                         Chanditala   
577                                           Nawabpur   
578                                          Bomgodaya   
579  Ghola, Housing Estate, Ramchandrapur, Tirthaba...   
580  Ghola, Housing Estate, Ramchandrapur, Tirthaba...   
581                           Barasat Bazar, Sethpukur   
582                           Barasat Bazar, Sethpukur   
583                           Barasat Bazar, Sethpukur   
584                           Barasat Bazar, Sethpukur   
585                           Barasat Bazar, Sethpukur   
586                                    B B Natun Bazar   
587                                    B B Natun Bazar   
588                                    B B Natun Bazar   
589                                    B B Natun Bazar   
590                                    B B Natun Bazar   
591                                    B B Natun Bazar   

                          Category   Distance   Latitude  Longitude PinCode  \
0                             Park   1.604059  22.573613  88.348341  700001   
1                    Boat or Ferry   1.604059  22.573613  88.348341  700001   
2                             None   5.219402  22.618940  88.373169  700002   
3                             Bank   3.482579  22.603822  88.367960  700003   
4                             Park   3.482579  22.6038

In [38]:
kolkata_20_df.to_csv('wb_area_pin_lat_lng_20km_venues.csv')

In [39]:
kolkata_20_df = pd.read_csv('wb_area_pin_lat_lng_20km_venues.csv', delimiter = ',')
kolkata_20_df = kolkata_20_df[[ 'PinCode', 'AreaName', 'Latitude', 'Longitude','Distance','VenueName','Category','VenueLatitude','VenueLongitude']]
kolkata_20_df

PinCode                                           AreaName   Latitude  \
0     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
1     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
2     700002  Cossipore Gun Factory, Cossipore H O, Indian R...  22.618940   
3     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
4     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
5     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
6     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
7     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
8     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
9     700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
10    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
11    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
12    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
13    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
14    700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...  22.598321   
15    700005                                Ahiritola, Hatkhola  22.598068   
16    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
17    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
18    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
19    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...  22.589073   
20    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
21    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
22    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
23    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
24    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
25    700008                             Barisha, Purba Barisha  22.481168   
26    700008                             Barisha, Purba Barisha  22.481168   
27    700009                                  Sarani Rarsibagan  22.576675   
28    700010          Beleghata H O, Joranandir, Subhas Sarobar  22.563026   
29    700010          Beleghata H O, Joranandir, Subhas Sarobar  22.563026   
..       ...                                                ...        ...   
562   712202                                            Mahes-I  22.735090   
563   712203                                        Mallickpara  22.745277   
564   712232                                         Bhadrakali  22.673991   
565   712233                                         Hind Motor  22.685856   
566   712234                                           Kanaipur  22.697519   
567   712235                              Cripar Road, Konnagar  22.700774   
568   712235                              Cripar Road, Konnagar  22.700774   
569   712245                                             Makhla  22.672289   
570   712246                                 Nabagram (Hooghly)  22.697671   
571   712247                             Raghunathpur (Hooghly)  22.657546   
572   712248  Bangurpark, Gandhisarak, Mahesh-Ii, Masjidlane...  22.723061   
573   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar  22.666507   
574   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar  22.666507   
575   712304                                              Janai  22.717670   
576   712702                                         Chanditala  22.696795   
577   712703                                           Nawabpur  22.696795   
578   743176                                          Bomgodaya  22.702216   
579   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...  22.697172   
580   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...  22.697172   
581   743201               

In [40]:
kolkata_20_park_df = kolkata_20_df[kolkata_20_df.Category == "Park"]
kolkata_20_park_df

PinCode                                           AreaName   Latitude  \
0     700001  Ananda Bazar Patrika, Braborne Road, Council H...  22.573613   
4     700003   Amrita Bazar, Bagh Bazar, Girish Avenue, Patrika  22.603822   
15    700005                                Ahiritola, Hatkhola  22.598068   
22    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
24    700007  Archana, Barabazar H O, College Squire, Kalake...  22.582557   
35    700012                             Bow Bazar, Sankaritala  22.568353   
55    700015              Seal Lane, Sales Tax Building, Tengra  22.555040   
102   700017  Circus Avenue, Jhowtala, Park Circus, Shakespe...  22.540940   
141   700020  Archarya Jagadish, Chandra Bose Road, Gokhale ...  22.537763   
150   700021                                       Fort Willium  22.563608   
168   700029  Dover Lane, Lake Market, Rash Behari Avenue, S...  22.517586   
290   700064                                       Bidhan Nagar  22.592457   
360   700073  Colootola, Chittaranjan Avenue, Kolkata Univer...  22.574311   
427   700091                                        Sech Bhavan  22.581696   
428   700091                                        Sech Bhavan  22.581696   

     Longitude  Distance                      VenueName Category  \
0    88.348341  1.604059                Millennium Park     Park   
4    88.367960  3.482579   Bagh Bazar Durga Puja Ground     Park   
15   88.358119  2.882242                 Kumartuli Park     Park   
22   88.361703  1.125577                    Girish Park     Park   
24   88.361703  1.125577      Victoria Memorial Grounds     Park   
35   88.361571  0.527732           Santosh Mitra Square     Park   
55   88.381247  2.639025                 Seal Lane Maat     Park   
102  88.368135  3.532956             Park Circus Maidan     Park   
141  88.351468  4.064120  Subhash Udyan (Northern Park)     Park   
150  88.338588  2.787156               Eden Garden Park     Park   
168  88.349433  6.271224                Deshapriya Park     Park   
290  88.412665  5.475682                  BF Block Park     Park   
360  88.361571  0.305403                   Md. Ali Park     Park   
427  88.430414  6.914019                      CKCL Park     Park   
428  88.430414  6.914019                  CK Block Park     Park   

     VenueLatitude  VenueLongitude  
0        22.572481       88.344517  
4        22.604344       88.366277  
15       22.598956       88.361514  
22       22.586069       88.362404  
24       22.585984       88.363124  
35       22.566008       88.365719  
55       22.551997       88.382552  
102      22.542981       88.368702  
141      22.535204       88.350350  
150      22.564244       88.340743  
168      22.518395       88.353177  
290      22.594503       88.414827  
360      22.577500       88.360720  
427      22.584534       88.427582  
428      22.584613       88.427625

In [41]:
kolkata_20_park_pins = kolkata_20_park_df['PinCode'].tolist()
kolkata_20_park_pins
kolkata_20_new_park_df = kolkata_20_df[["PinCode","AreaName"]].drop_duplicates()
kolkata_20_new_park_df = kolkata_20_new_park_df[~kolkata_20_new_park_df["PinCode"].isin(
kolkata_20_park_pins)]
kolkata_20_new_park_df

PinCode                                           AreaName
2     700002  Cossipore Gun Factory, Cossipore H O, Indian R...
9     700004  R G Kar M C Hospital, Shyambazar, Shyambazar M...
16    700006  Beadon Street, Bidhan Sarani, Jorasanko, Manik...
25    700008                             Barisha, Purba Barisha
27    700009                                  Sarani Rarsibagan
28    700010          Beleghata H O, Joranandir, Subhas Sarobar
31    700011                                        Narkeldanga
36    700013                              Dharamtala, Mott Lane
51    700014  Bamboovilla, Intally, Linton Street, Sealdah, ...
57    700016             Elliot Road, Madrassa, Park Street H O
109   700018                             Bartala, Rabindranagar
110   700019  Ballygunge, Ballygunge R S, Ballygunge Science...
151   700022                              Bakery Road, Hastings
152   700023  Kidderpore, Manstala, Market, Raja Janakinath,...
153   700024  Garden Reach, Keshoram Cotton Mills, Paharpur ...
154   700025  Bhawanipore, Lansdown Market, Ranshet Wadi, Ra...
155   700026  Keoratala, Kalighat, Kalimandir, Protisthan, R...
159   700027  Alipore Bodyguard Line, Alipore Civil Court, A...
164   700028  Dum Dum, Jungipara Salgs Nager Bzr, Jugipara S...
179   700030              Ghughudanga, Purbe Sinthee, Sethbagan
183   700031                             Dhakuria, K P Roy Lane
188   700032  Bejoy Garh, Hossain Road, Jadavpur University,...
193   700033  Indrani Park, Tallygunge H O, Tollygunge Circu...
197   700034  Brahma Samaj Road, Behala, Behala Municipal Ma...
202   700035  Alambazar, Ashokegarh, Bon Hooghly, Desh Bandh...
206   700036                         Bengal Immunity, Baranagar
209   700037         Belgachia, Belgachia Road, Northern Avenue
212   700038                                            Sahapur
215   700039  Bediadanga, Picnic Garden, Topsia, Tiljala, Ti...
223   700040                    Netajinagar, Regent Park, Russa
..       ...                                                ...
547   711205                                      Sambay Pallay
549   711227                                          Ghoshpara
550   711302                                        Andul Mauri
552   711309                                        Delta Mills
553   711311                               Jagaeha, Old Jagacha
554   711313                                           Sankrail
555   711402                                          Baltikuri
556   711403                                               Bank
557   711405                                             Domjur
558   711409                                           Makardah
559   711411                                              Begri
560   712201  Kalbazar, Manicktala Bazar, Serampore Court, S...
562   712202                                            Mahes-I
563   712203                                        Mallickpara
564   712232                                         Bhadrakali
565   712233                                         Hind Motor
566   712234                                           Kanaipur
567   712235                              Cripar Road, Konnagar
569   712245                                             Makhla
570   712246                                 Nabagram (Hooghly)
571   712247                             Raghunathpur (Hooghly)
572   712248  Bangurpark, Gandhisarak, Mahesh-Ii, Masjidlane...
573   712258        Raiendra Avenue, Uttarpara, Uttarpara Bazar
575   712304                                              Janai
576   712702                                         Chanditala
577   712703                                           Nawabpur
578   743176                                          Bomgodaya
579   743178  Ghola, Housing Estate, Ramchandrapur, Tirthaba...
581   743201                           Barasat Bazar, Sethpukur
586   743319                                    B B Natun Bazar

[173 rows x 2 columns]

### So there are 171 new Children Parks will be recommended to the government. The above location information will be handed over to the Kolkata Corporation.